Wrap future continuations as well

taion opened the pull request:
n.b. I've based this PR on the `fix-instanceof` branch because it depends on #39 

The new commit fixes the issue I reported at https://github.com/othiym23/node-continuation-local-storage/issues/41 where chained continuations weren't appropriately wrapped up here.

hayes commented:
Thanks for working on getting this fixed!

This looks like it might be the right approach.  Would you mind splitting the tests that test this specific issue into separate tests? Modifying existing tests makes it hard to tell what tests are changing because behavior is changing and what tests are changing because the test case is being expanded (does that make sense)

taion commented:
Thank you - I almost gave up and just patched CLS.

Good point on the tests, I'll make that change.

taion commented:
Updated and squashed. Also changed the accept/reject tests to be a bit more parallel to the rest of the tests to show that the continuations were getting set up correctly there as well.

hayes commented:
Thank you for updating this, the commit diff is much easier to read now.  Looks like there is some confusing stuff happening around `Promise.all` and `Promise.race` which appear to be creating some additional contexts internally resulting in duplicated names.  Not sure how to address this, maybe just comments explaining why the have a name, or add a way to queue up values for `currentName` so that a call that generates multiple calls can consume from that queue (thats probably over engineered and a terrible idea, just thinking out loud)  Tests for this stuff are really frustrating to write

taion commented:
Yes, there's definitely confusing stuff going on in `all` and `race`. I agree with your diagnosis of what's happening there. I'll take another look at making the output there less confusing.

Definitely agree that these sorts of tests are finicky to write - it's pretty surprising by comparison how the tests I care about at a higher level ended up being so much easier: https://github.com/othiym23/node-continuation-local-storage/pull/42

taion commented:
BTW, that's exactly what's happening with `all` and `race` (`PromiseAll` and `PromiseOne` here respectively): https://github.com/joyent/node/blob/d13d7f74d794340ac5e126cfb4ce507fe0f803d5/deps/v8/src/promise.js#L273-L323

taion commented:
I couldn't find a good way to get a better `currentName` without doing something that depended too much on the internal implementations of `PromiseOne` and `PromiseAll`, so I just added comments to the expected values to describe what those extra nodes were.

hayes commented:
looks like my PR got merged, might want to rebase onto master.  Looks good to me though! Need to get @othiym23 to merge and release

taion commented:
Rebased onto the new master.

hayes commented:
Reading through the v8 source would have been a great idea when I first wrote this (sorry it was buggy)

taion commented:
I can honestly tell you that for the most part, it only served to confuse me, and accomplished very little beyond wasting a lot of my time chasing dead ends before I realized that I could do what I wanted in pretty much a (hopefully pretty straightforward) one-liner.

hayes commented:
Sorry I didn't notice this earlier, but was thinking about your solution, and I believe it will cause some issues. 

```javascript
Promise.accept().then(function delayedAccept(accept) {
  return new Promise(function wait(accept) {
    setTimeout(function resolve() {
      accept(10)
    })
  })
}).then(function whereWasIContinuedFrom() {
  // should be continued from setTimeout
  // will be continued from then(delayedAccept)
})
```

Have not yet tested this, but looking at it it looks like anything continued from a then will be a direct continuation from that `then` rather than the correct child of `then`.  Maybe this was a flawed assumption, but the Promise instrumentation was originally written to continue from an `accept` or `reject` call to the callback to `then`, `catch` or `chain` rather than continuing from call to `then` to `then` callback.

taion commented:
I see what you mean. It is as you describe. That said, the current version actually will have `whereWasIContinuedFrom` run as a continuation of nothing at all, which is worse than this behavior. I agree that what you propose would be ideal, though.

taion commented:
I'm getting no love from this. If I run something like

```js
import 'continuation-local-storage';

Promise.resolve()
  .then(() => {
    Promise.resolve()
      .then(() => Promise.resolve())
      .then(() => Promise.resolve());
  })
  .then(() => {});
```

and add some logging to `wrappedThen`, I expect to see per https://github.com/joyent/node/blob/885f72105bf0b18007100550f131c85ca56536e0/deps/v8/src/promise.js#L259 that `wrappedThen` will get called maybe 7 or 8 times (i.e. when a promise resolves with another promise, then every continuation of that promise should get `.then()` called on it again).

But I don't see this at all, so I have no hook to replace the `__asl_wrapper` for the continuation.

I'm aware that the change as it stands is not entirely ideal in this respect. However, I believe in the scenario you outlined, running as a continuation of `wait` is "less wrong" than running as a continuation of nothing at all. However, I really don't see a way forward in this respect.

hayes commented:
I spent a bit of time trying to solve this for the newrelic agent, and I think I have a solution that will work.  I may send a PR to your branch

hayes commented:
Sent a PR to your branch (https://github.com/taion/async-listener/pull/1/files) tests need to be fixed so they also run on iojs@1.0.* but I think the implementation should work

taion commented:
That's really cool. I was about to get started on writing a much less clean version of that. Will add comments on that PR.

hayes commented:
I figured out the issue I was running into, and why I was creating the wrapper before the function (I didn't know the reason, just that there was an issue with waiting until after).  Issue is that the function passed to a continuation can throw, so naively creating a wrapper after does not guarantee that it will actually happen, a try-finally should do the trick, PR incoming

taion commented:
I can add that - I'm mostly done updating the new test cases for changing back where the wrapper was getting created.

taion commented:
Do the new tests only fail on iojs 1.0, or will I see the failures on iojs 2.0 as well?

hayes commented:
The tests I wrote all passed on iojs 2.0 but there were a few failures on 1.0, didn't check 0.12 but it should be the same as 1.0

hayes commented:
if you are working on 0.12 should work for 1.0 I can fix for 2.0, basically 2.0 changed the implementation slightly and creates some extra async contexts

hayes commented:
I am happy to help get tests passing

taion commented:
I only asked because I'm feeling too lazy to figure out how to install iojs 1.x with Homebrew (:

The iojs 2.x test failures (maybe that I just introduced back again) seemed pretty straightforward to fix in line with the earlier ones, so I just went and fixed them.

I'd be glad for any help in getting tests passing on any platforms that I missed. Thanks!

hayes commented on index.js:
```diff
@@ -304,23 +302,50 @@ if (instrumentPromise) {
+      return nextResult;
+    } finally {
+      // Wrap any resulting futures as continuations.
+      if (nextResult instanceof wrappedPromise) {
```
so this is kind of a dumb edge case, but I had a discussion earlier with a coworker about what this instanceof check should be checking against.  I believe it would be better to use the original promise incase you pass an unwrapped promise from to then. Your not going to get the best results if your mixing unwrapped promises into things, but for the sake of interoperability with other monkey patching I think its best to leave the most basic check possible here.

taion replied:
Makes sense. I misunderstood what was happening. I'll fix it.

hayes commented on index.js:
```diff
@@ -304,23 +302,50 @@ if (instrumentPromise) {
+      // Wrap any resulting futures as continuations.
+      if (nextResult instanceof wrappedPromise) {
+        next.__asl_wrapper = function proxyWrapper() {
+          var aslWrapper = nextResult.__asl_wrapper || propagateAslWrapper;
```
Ah good change not using my dumb defaultWrapper solution here :+1:

hayes commented:
This looks good to me
@othiym23 if you could take a look/merge this when you have some time that would be awesome!

hayes commented:
Awesome work on this @taion very happy to have someone else work on this stuff, I am not a Promise user, so its good to get a second set of eyes on this and get things polished up.

taion commented:
@hayes Thank you! I've almost got something really cool working with continuation-local-storage, but hit this little bit of a road block here since all of my async was done with promises, and it was quite a fun little problem to work on.

@othiym23 If this is good to merge, I'd appreciate it if you could also merge the tests covering promises in the CLS package and cut a new release that bumps the async-listener dependency.

othiym23 commented:
Thanks to both of you for patiently working through this. The lack of observability of the microtask queue from the Node side of things is a serious pain in the butt, and I'm guessing we'll probably see more issues around this, but this is a good patch to have in place. Landed in fad6f737d72320e4cb638a4a9b9adfb0949be833 and others.

taion commented:
@othiym23 

Thank you! Hopefully most of the edge cases have been resolved now, but there's probably some room to improve this code regardless with things like not keeping around as many references to old promises to handle promise unwrapping in `then`.

Would you be able to look at https://github.com/othiym23/node-continuation-local-storage/pull/42 and possibly merge and cut a new release there with the new async-listener dependency? This (admittedly simpler) use case was what motivated my looking at this in the first place, and is a bit more approachable w/r/t what's going on.
